### PR TITLE
Assets pallet: Giving the asset owner the ability to set minimum balance

### PR DESCRIPTION
**Pull request by Szegoo**
Closes: #13402

Introduces a new extrinsic `set_min_balance` which allows the owner of the specific asset to change the `min_balance` required for an asset.

This is possible to do if there aren't any accounts holding the asset or if the new min balance is less than the old min balance.

Cumulus companion: https://github.com/paritytech/cumulus/pull/2253

### CODE REVIEW

File: frame/assets/src/lib.rs [hunk 1]
```diff
@@ -1511,6 +1511,45 @@ pub mod pallet {
...
+		///
+		/// Origin must be Signed and the sender has to be the Owner of the
+		/// asset `id`.
+		///
+		/// - `id`: The identifier of the asset.
+		/// - `min_balance`: The new value of `min_balance`.
+		///
+		/// Emits `AssetStatusChanged` event when successful.
+		#[pallet::call_index(28)]
+		#[pallet::weight(10_000)]
```
Comment (Szegoo) [inline frame/assets/src/lib.rs:1527]: Will update this after weight generation.

**Comment by Szegoo** [comment]
@jsidorenko @bkchr Could you review this?

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -1511,6 +1511,45 @@ pub mod pallet {
...
+			// any accounts holding the asset.
+			ensure!(
+				min_balance < old_min_balance || details.accounts == 0,
+				Error::<T, I>::NoPermission
+			);
+
+			details.min_balance = min_balance;
+			Asset::<T, I>::insert(&id, details);
+
+			Self::deposit_event(Event::AssetStatusChanged { asset_id: id });
```
Comment (jsidorenko) [inline frame/assets/src/lib.rs:1550]: Shall we emit smth more specific? Like AssetMinBalanceChanged and pass the new value into the event data?

Comment (Szegoo) [reply frame/assets/src/lib.rs:1550]: Makes sense, I will add that.

**Review by jsidorenko — APPROVED** [review]

File: frame/assets/src/lib.rs [hunk 3]
```diff
@@ -1511,6 +1513,48 @@ pub mod pallet {
...
+		/// accounts that are holding the asset or if the new vlaue of new_balance
+		/// is less than the old one.
+		///
+		/// Origin must be Signed and the sender has to be the Owner of the
+		/// asset `id`.
+		///
+		/// - `id`: The identifier of the asset.
+		/// - `min_balance`: The new value of `min_balance`.
+		///
+		/// Emits `AssetStatusChanged` event when successful.
```
Comment (bkchr) [inline frame/assets/src/lib.rs:1527]:
```suggestion
		/// Emits `AssetMinBalanceChanged` event when successful.
```

File: frame/assets/src/lib.rs [hunk 4]
```diff
@@ -1511,6 +1513,48 @@ pub mod pallet {
 			let id: T::AssetId = id.into();
 			Self::do_refund(id, ensure_signed(origin)?, allow_burn)
 		}
+
+		/// Sets the minimum balance of an asset. Only works if there aren't any
+		/// accounts that are holding the asset or if the new vlaue of new_balance
+		/// is less than the old one.
```
Comment (bkchr) [inline frame/assets/src/lib.rs:1519]:
```suggestion
		/// Sets the minimum balance of an asset.
		///
		/// Only works if there aren't any accounts that are holding the asset or if 
		/// the new value of `min_balance` is less than the old one.
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
@ggwpez could you please summon the mighty benchmark bot?

**Comment by ggwpez** [comment]
bot bench $ pallet dev pallet_assets

**Comment by ggwpez** [comment]
bot help

**Comment by jsidorenko** [comment]
bot merge

**Comment by IkerAlus** [comment]
Sorry for jumping in when this is already merged but I want to make sure this new extrinsic is not opening a door for blockchain state spamming:

Image the following: An asset is created on Statemint, the team behind the asset makes a gov proposal to make it sufficient with a reasonable `min_balance` similar to the ED in DOTs. The gov proposal is passed, then the team behind the asset (owner of the asset) lowers down the `min_balance` via this new extrinsic to a ridiculously small value. Attackers start spamming the blockchain with millions of super cheap transfer transactions to random accounts (fees will be low since they are proportional to `min_balance` value).

How is this scenario prevented? I was expecting to see a check for asset sufficiency to stop the update of the `min_balance` via the extrinsic.

**Comment by jsidorenko** [comment]
@IkerAlus You're right! There should be a check for sufficiency in that call, as you'be described in the issue

**Pull request merged**